tools/stimulus.py

```diff
@@ -163,152 +181,162 @@ def main():
         elif o == "-t":
             types.append(a)
         elif o == "-f":
-            functions.append(a)        
+            functions.append(a)
         elif o == "-l":
             languages.append(a)
         elif o == "-i":
             inputs.append(a)
-        elif o =="-v":
-            verbose=True
+        elif o == "-v":
+            verbose = True
 
     # Parameter checks
     # Note: the lists might be empty, but languages and outputs must
     # have the same length.
     if len(languages) != len(outputs):
-        print "Error: number of languages and output files must match"
+        print("Error: number of languages and output files must match")
         sys.exit(4)
     for l in languages:
-        if not l+"CodeGenerator" in globals():
-            print "Error: unknown language:", l
+        if not l + "CodeGenerator" in globals():
+            print("Error: unknown language:", l)
             sys.exit(6)
     for f in types:
         if not os.access(f, os.R_OK):
-            print "Error: cannot open type file:", f
+            print("Error: cannot open type file:", f)
             sys.exit(5)
     for f in functions:
         if not os.access(f, os.R_OK):
-            print "Error: cannot open function file:", f
+            print("Error: cannot open function file:", f)
             sys.exit(5)
     for f in inputs:
         if not os.access(f, os.R_OK):
-            print "Error: cannot open input file:", f
+            print("Error: cannot open input file:", f)
             sys.exit(5)
     # TODO: output files are not checked now
 
     # OK, do the trick:
-    for l in range(len(languages)):
-        cl=globals()[languages[l]+"CodeGenerator"]
-        cg=cl(functions, types)
-        cg.generate(inputs, outputs[l])
+    for language, output in zip(languages, outputs):
+        cl = globals()[language + "CodeGenerator"]
+        cg = cl(functions, types)
+        cg.generate(inputs, output)
+
 
 ################################################################################
-class CodeGenerator:
+class CodeGenerator(metaclass=abc.ABCMeta):
     def __init__(self, func, types):
-        # Set name
-        self.name=str(self.__class__).split(".")[-1]
-        self.name=self.name[0:len(self.name)-len("CodeGenerator")]
+        # Set name, note this only works correctly if derived classes always
+        # extend it as by prepending the language to the CodeGenerator class
+        # name
+        self.name = type(self).__name__
+        self.name = self.name[0 : len(self.name) - len("CodeGenerator")]
 
         # Parse function and type files
-        parser=PParser()
-        self.func=seqdict.seqdict()
+        parser = PParser()
+        self.func = OrderedDict()
         for f in func:
-            ff=open(f)
-            newfunc=parser.parse(ff)
-            self.func.extend(newfunc)
+            ff = open(f)
```

Comment:
`open()` is mostly used nowadays as a context manager, like:

```
with open(f) as ff:
    ...do something with ff here, close() is called automatically when exiting the context...
```

This is not urgent either, just something for the future.